crypto/rsa/rsa_ossl.c

```diff
@@ -664,10 +668,37 @@ static int rsa_ossl_mod_exp(BIGNUM *r0, const BIGNUM *I, RSA *rsa, BN_CTX *ctx)
     }
 
     if (rsa->flags & RSA_FLAG_CACHE_PUBLIC)
-        if (!BN_MONT_CTX_set_locked
-            (&rsa->_method_mod_n, rsa->lock, rsa->n, ctx))
+        if (!BN_MONT_CTX_set_locked(&rsa->_method_mod_n, rsa->lock,
+                                    rsa->n, ctx))
+            goto err;
+
+    if (smooth) {
+        if (/* m1 = I moq q */
+            !bn_from_mont_fixed_top(m1, I, rsa->_method_mod_q, ctx)
+            || !bn_to_mont_fixed_top(m1, m1, rsa->_method_mod_q, ctx)
+            /* m1 = m1^dmq1 mod q */
+            || !BN_mod_exp_mont_consttime(m1, m1, rsa->dmq1, rsa->q, ctx,
+                                          rsa->_method_mod_q)
+            /* r1 = I mod p */
+            || !bn_from_mont_fixed_top(r1, I, rsa->_method_mod_p, ctx)
+            || !bn_to_mont_fixed_top(r1, r1, rsa->_method_mod_p, ctx)
+            /* r1 = r1^dmp1 mod p */
+            || !BN_mod_exp_mont_consttime(r1, r1, rsa->dmp1, rsa->p, ctx,
+                                          rsa->_method_mod_p)
+            /* r1 = (r1 - m1) mod p */
+            || !bn_mod_sub_fixed_top(r1, r1, m1, rsa->p)
+
+            /* r0 = r0 * iqmp mod p */
+            || !bn_to_mont_fixed_top(r1, r1, rsa->_method_mod_p, ctx)
+            || !bn_mul_mont_fixed_top(r1, r1, rsa->iqmp, rsa->_method_mod_p,
+                                      ctx)
+            || !bn_mul_fixed_top(r0, r1, rsa->q, ctx)
+            || !bn_mod_add_fixed_top(r0, r0, m1, rsa->n))
```

Comment:
`r0` here is fixed-top and yet leaks out of the public API, by way of `RSA_meth_get_mod_exp` or 1.0.x's public structs. Even separate from the public API, it leaks into `BN_sub` via the unblinded `RSA_X931_PADDING` path of `rsa_ossl_private_encrypt`. It leaks into `BN_bn2binpad` when unblinded which does not actually tolerate non-minimal things yet. (It calls `BN_num_bytes` and the like.) It also leaks into `BN_BLINDING_invert_ex` which currently uses plain `BN_mod_mul` (though we want Montgomery reduction). I *think* that tolerates it because `BN_mul` and `BN_sqr` calls `bn_correct_top` (though we don't want a `bn_correct_top` in there), but that's a tenuous set of requirements...

This is unfortunate since it is indeed one of the places where we *don't* want to force minimal-width `BIGNUM`s, but that may require a more complete solution to #6640.

Reply:
> This is unfortunate since it is indeed one of the places where we don't want to force minimal-width BIGNUMs, but that may require a more complete solution to #6640.

(I guess, when blinding is on, we're vaguely okay with forcing minimal-width here. It's the blinding removal that's sensitive. Anyway, point still stands that we probably need to force minimal-width for now, but we don't want to long-term.)

Reply:
Looking into blinding is the tomorrow's exercise...